### PR TITLE
Fix possible race condition in ProjectRootElementCache.Get

### PR DESCRIPTION
**Pull request by AR-May**
Fixes possible race condition in `ProjectRootElementCache.Get` supposedly introduced in #6680.

### Context
Passing a load function that uses `ProjectRootElementCache.Get` to `ProjectRootElementCache.Get` function could lead to failure (to verify a path in the loaded ProjectRootElement), due to a race condition. We eliminate calling `ProjectRootElementCache.Get` inside the load function, eliminating an unnecessary double entry to `ProjectRootElementCache.Get` function.

### Changes Made
- Remove double entry to `ProjectRootElementCache.Get` function that could cause race condition.
- Adjust path verification in `ProjectRootElementCache` and `SimpleProjectRootElementCache` to be more safe and to have more information.

### Testing
Unit tests + manual testing

### Notes
The stack trace of failure:

```
> ##[error]C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\CommonExtensions\Microsoft\NuGet\NuGet.RestoreEx.targets(19,5): Error : MSB0001: Internal MSBuild Error: Got project back with incorrect path 
> at Microsoft.Build.Shared.ErrorUtilities.ThrowInternalError(String message, Exception innerException, Object[] args) 
> at Microsoft.Build.Evaluation.ProjectRootElementCache.Get(String projectFile, OpenProjectRootElement openProjectRootElement, Boolean isExplicitlyLoaded, Nullable1 preserveFormatting) 
> at Microsoft.Build.Evaluation.Evaluator4.ExpandAndLoadImportsFromUnescapedImportExpression(String directoryOfImportingFile, ProjectImportElement importElement, String unescapedExpression, Boolean throwOnFileNotExistsError, List1& imports) 
> at Microsoft.Build.Evaluation.Evaluator4.ExpandAndLoadImportsFromUnescapedImportExpressionConditioned(String directoryOfImportingFile, ProjectImportElement importElement, List1& projects, SdkResult& sdkResult, Boolean throwOnFileNotExistsError) 
> at Microsoft.Build.Evaluation.Evaluator4.ExpandAndLoadImports(String directoryOfImportingFile, ProjectImportElement importElement, SdkResult& sdkResult) 
> at Microsoft.Build.Evaluation.Evaluator4.EvaluateImportElement(String directoryOfImportingFile, ProjectImportElement importElement) 
> at Microsoft.Build.Evaluation.Evaluator4.PerformDepthFirstPass(ProjectRootElement currentProjectOrImport) 
> at Microsoft.Build.Evaluation.Evaluator4.EvaluateImportElement(String directoryOfImportingFile, ProjectImportElement importElement) 
> at Microsoft.Build.Evaluation.Evaluator4.PerformDepthFirstPass(ProjectRootElement currentProjectOrImport) 
> at Microsoft.Build.Evaluation.Evaluator4.EvaluateImportElement(String directoryOfImportingFile, ProjectImportElement importElement) 
> at Microsoft.Build.Evaluation.Evaluator4.PerformDepthFirstPass(ProjectRootElement currentProjectOrImport) 
> at Microsoft.Build.Evaluation.Evaluator4.EvaluateImportElement(String directoryOfImportingFile, ProjectImportElement importElement) 
> at Microsoft.Build.Evaluation.Evaluator4.PerformDepthFirstPass(ProjectRootElement currentProjectOrImport) 
> at Microsoft.Build.Evaluation.Evaluator4.Evaluate() 

```

### CODE REVIEW

File: src/Build/Evaluation/ProjectRootElementCache.cs [hunk 1]
```diff
@@ -275,7 +275,13 @@ private bool IsInvalidEntry(string projectFile, ProjectRootElement projectRootEl
 
                 projectRootElement = openProjectRootElement(projectFile, this);
                 ErrorUtilities.VerifyThrowInternalNull(projectRootElement, "projectRootElement");
-                ErrorUtilities.VerifyThrow(projectRootElement.FullPath == projectFile, "Got project back with incorrect path");
+                ErrorUtilities.VerifyThrow(
```
Comment (AR-May) [inline src/Build/Evaluation/ProjectRootElementCache.cs:278]: This line basically checks if `projectFile` that came to this function is indeed a full path, as it should be.

File: src/Build/Evaluation/ProjectRootElementCache.cs [hunk 2]
```diff
@@ -275,7 +275,13 @@ private bool IsInvalidEntry(string projectFile, ProjectRootElement projectRootEl
...
+                    projectRootElement.FullPath.Equals(projectFile, StringComparison.OrdinalIgnoreCase),
+                    "Got project back with incorrect path. Expected path: {0}, received path: {1}.",
+                    projectFile,
+                    projectRootElement.FullPath
+                );
+                //ErrorUtilities.VerifyThrow(projectRootElement.FullPath == projectFile, "Got project back with incorrect path, first path = {1} {}");
```
Comment (Forgind) [inline src/Build/Evaluation/ProjectRootElementCache.cs:284]: nit:
delete commented out code

**Review by Forgind — COMMENTED** [review]
I imagine this would hurt perf? Do you have numbers on by how much?

File: src/Build/Evaluation/ProjectRootElementCache.cs (hunk 2 above)
Comment (AR-May) [reply src/Build/Evaluation/ProjectRootElementCache.cs:284]: oh, right

**Comment by rainersigwald** [comment]
Would you mind pasting a stack that shows the double call into the PR description? I remember this looked good when we talked about it but I've forgotten the details and they'd be good to have here for posterity.

**Comment by AR-May** [comment]
No, it should not affect perf at all. Posted a stack trace of the error in the description.

The problem here is that we had double entry into Get() function of a cache, which is completely unnecessary and under some conditions it may lead to modifying `FullPath` right after line 276 in `ProjectRootElementCache.cs` and before the check on line 278. Logic is following: If load function `openProjectRootElement` has a Get() call inside, we do not load the xml and instead take element from a cache. 'projectRootElement ' is no more only locally accessible variable and some other thread could change it. Therefore, the check fails.

**Comment by AR-May** [comment]
The stack trace does not show the double call, cause the error happens in the time when we are already in the first call and just came out of the second call.

**Review by rokonec — APPROVED** [review]

**Comment by rokonec** [comment]
@rainersigwald 

> Would you mind pasting a stack that shows the double call into the PR description? I remember this looked good when we talked about it but I've forgotten the details and they'd be good to have here for posterity.

It was OK, but recently we have made changes that `Get` was not in one atomic lock but split into multiple lock se we can run loading of XML outside of lock, and this is sensitive against Get same thread reentrancy/recursion.

**Review by Forgind — APPROVED** [review]

**Review by ladipro — APPROVED** [review]

File: src/Build/Evaluation/ProjectRootElementCache.cs [hunk 3]
```diff
@@ -209,13 +218,18 @@ private bool IsInvalidEntry(string projectFile, ProjectRootElement projectRootEl
...
+        /// <param name="loadProjectRootElement">The delegate to use to load if necessary. May be null. Must not update the cache.</param>
         /// <param name="isExplicitlyLoaded"><code>true</code> if the project is explicitly loaded, otherwise <code>false</code>.</param>
         /// <param name="preserveFormatting"><code>true</code> to the project was loaded with the formated preserved, otherwise <code>false</code>.</param>
         /// <returns>The ProjectRootElement instance if one exists.  Null otherwise.</returns>
-        internal override ProjectRootElement Get(string projectFile, OpenProjectRootElement openProjectRootElement, bool isExplicitlyLoaded,
+        internal override ProjectRootElement Get(string projectFile, OpenProjectRootElement loadProjectRootElement, bool isExplicitlyLoaded,
             bool? preserveFormatting)
         {
+#if DEBUG
+            s_getEntriesNumber++;
```
Comment (KirillOsenkov) [inline src/Build/Evaluation/ProjectRootElementCache.cs:229]: `Interlocked.Increment(ref s_getEntriesNumber)` perhaps?

File: src/Build/Evaluation/ProjectRootElementCache.cs [hunk 4]
```diff
@@ -253,15 +267,21 @@ private bool IsInvalidEntry(string projectFile, ProjectRootElement projectRootEl
                 ForgetEntryIfExists(projectRootElement);
             }
 
-            if (openProjectRootElement == null)
+            if (loadProjectRootElement == null)
             {
                 if (projectRootElement == null || projectRootElementIsInvalid)
                 {
+#if DEBUG
+                    s_getEntriesNumber--;
```
Comment (KirillOsenkov) [inline src/Build/Evaluation/ProjectRootElementCache.cs:275]: `Interlocked.Decrement`

File: src/Build/Evaluation/ProjectRootElementCache.cs (hunk 3 above)
Comment (AR-May) [reply src/Build/Evaluation/ProjectRootElementCache.cs:229]: Yes, indeed. Forgot to return this back on my way while was removing and adding back the thread static modifier.

**Review by rokonec — APPROVED** [review]

File: src/Build/Evaluation/ProjectRootElementCache.cs [hunk 5]
```diff
@@ -209,92 +222,115 @@ private bool IsInvalidEntry(string projectFile, ProjectRootElement projectRootEl
...
+        internal override ProjectRootElement Get(string projectFile, OpenProjectRootElement loadProjectRootElement, bool isExplicitlyLoaded,
             bool? preserveFormatting)
         {
-            // Should already have been canonicalized
-            ErrorUtilities.VerifyThrowInternalRooted(projectFile);
+#if DEBUG
+            // Verify that loadProjectRootElement delegate does not call ProjectRootElementCache.Get().
+            Interlocked.Increment(ref s_getEntriesNumber);
+            ErrorUtilities.VerifyThrow(
+                s_getEntriesNumber == 1,
```
Comment (rainersigwald) [inline src/Build/Evaluation/ProjectRootElementCache.cs:236]: Should you be verifying the return value of `Interlocked.Increment` instead?

**Review by rainersigwald — APPROVED** [review]

File: src/Build/Evaluation/ProjectRootElementCache.cs (hunk 5 above)
Comment (KirillOsenkov) [reply src/Build/Evaluation/ProjectRootElementCache.cs:236]: Good point, there’s a window between the increment and the check where the value could have changed again.

Comment (AR-May) [reply src/Build/Evaluation/ProjectRootElementCache.cs:236]: Hmm, it seems i was a bit confused when I changed to `Interlocked.Increment`. We do not need it, cause the value `s_getEntriesNumber` is thread static, not just static. Our goal is to catch if we have two `Get()` calls in the call stack (possibly from different instances of `ProjectRootElementCache`). So, there is no chance that value would be changed from some other thread. I will return back to usual increments/decrements here.

Comment (KirillOsenkov) [reply src/Build/Evaluation/ProjectRootElementCache.cs:232]: makes sense

**Pull request merged**